keyman_trust_linux.go

```diff
@@ -4,71 +4,90 @@ import (
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("Unable to run certutil command: %s\n%s", err, out)
+		}
+		return nil
+	})
 }
 
 // AddAsTrustedRoot adds the certificate to the user's trust store as a trusted
-// root CA.
-// Note - on Linux, this assumes the user is using Chrome.
+// root CA. Supports Chrome and Firefox
 func (cert *Certificate) AddAsTrustedRoot(prompt string) error {
 	tempFileName, err := cert.WriteToTempFile()
 	defer os.Remove(tempFileName)
 	if err != nil {
 		return fmt.Errorf("Unable to create temp file: %s", err)
 	}
 
-	nssdb, err := getUserNssdb()
-	if err != nil {
-		return err
-	}
-
-	// Add it as a trusted cert
-	// https://code.google.com/p/chromium/wiki/LinuxCertManagement#Add_a_certificate
-	cmd := exec.Command("certutil", "-d", nssdb, "-A", "-t", "C,,", "-n", cert.X509().Subject.CommonName, "-i", tempFileName)
-	out, err := cmd.CombinedOutput()
-	if err != nil {
-		return fmt.Errorf("Unable to run certutil command: %s\n%s", err, out)
-	} else {
+	return forEachNSSProfile(func(profile string) error {
+		// Add it as a trusted cert
+		// https://code.google.com/p/chromium/wiki/LinuxCertManagement#Add_a_certificate
+		cmd := exec.Command("certutil", "-d", profile, "-A", "-t", "C,,", "-n", cert.X509().Subject.CommonName, "-i", tempFileName)
+		out, err := cmd.CombinedOutput()
+		if err != nil {
+			return fmt.Errorf("Unable to run certutil command: %s\n%s", err, out)
+		}
 		return nil
+	})
+}
+
+func pathExists(path string) bool {
+	_, err := os.Stat(path)
+	return err == nil
+}
+
+func forEachNSSProfile(f func(profile string) error) error {
+	profiles, _ := filepath.Glob(FirefoxProfile)
+	profiles = append(profiles, nssDBs...)
+	for _, profile := range profiles {
+		if stat, err := os.Stat(profile); err != nil || !stat.IsDir() {
+			continue
+		}
+		if pathExists(filepath.Join(profile, "cert9.db")) {
+			if err := f("sql:" + profile); err != nil {
+				return err
+			}
+		} else if pathExists(filepath.Join(profile, "cert8.db")) {
+			if err := f("dbm:" + profile); err != nil {
+				return err
+			}
+		}
 	}
+	return nil
 }
 
-// Checks whether this certificate is install based purely on looking for a cert
+// IsInstalled checks whether this certificate is install based purely on looking for a cert
 // in the user's nssdb that has the same common name.  This function returns
 // true if there are one or more certs in the nssdb whose common name
 // matches this cert.
 func (cert *Certificate) IsInstalled() (bool, error) {
-	nssdb, err := getUserNssdb()
-	if err != nil {
-		return false, err
-	}
-
-	cmd := exec.Command("certutil", "-d", nssdb, "-L", "-n", cert.X509().Subject.CommonName)
-	err = cmd.Run()
+	found := false
+	err := forEachNSSProfile(func(profile string) error {
+		cmd := exec.Command("certutil", "-d", profile, "-L", "-n", cert.X509().Subject.CommonName)
+		err := cmd.Run()
 
-	found := err == nil
-	return found, nil
-}
+		if err == nil {
```

Comment:
This is a little tricky. IIRC the use of IsInstalled is to check whether the certificate is installed and to not bother prompting the user to install it if it already is.
Since we're not dealing with multiple locations where the certificate may or may not be installed, I think we need to actually check and install each location individually. So for example, if you're already got the certificate installed for Chrome, but you don't have it installed for Firefox, we should prompt you to install it for Firefox.

This means that the public API might need to change. Instead of having both `IsInstalled` and `AddAsTrustedRoot`, I think we might need a `AddAsTrustedRootIfNecessary` or something like that.

Reply:
@oxtoacart I'm not sure I see the issue.
IsInstalled now checks if the cert is installed in any keyring either chrome or firefox.
AddAsTrusted adds where it can.
When does this cause a problem?

Reply:
1. User is running Chrome
2. User starts Lantern
3. Lantern calls IsInstalled to see if the cert is installed, which returns false
4. Lantern calls AddAsTrustedRoot which installs the cert for Chrome
5. User decides she likes Firefox better, installs Firefox
6. User runs Lantern
7. Lantern calls IsInstalled to see if the cert is installed, which returns true
8. The user browses in Firefox, but Lantern cannot MITM because Firefox doesn't have the cert installed

Reply:
@oxtoacart that makes sense.
How would you envision the API change? A method that receives the name of the browser? Or just get rid of IsInstalled and always try to install where you can (if needed)?

Reply:
[Here](https://github.com/getlantern/mitm/blob/master/mitm.go#L188) is the logic that currently uses these functions. It might be best just to provide an `InstallCertIfNecessary` function in keyman which we can call from mitm. We might also need to pull [this Windows-specific logic](https://github.com/getlantern/mitm/blob/master/mitm.go#L197) into keyman.

Reply:
One thing to make sure of, if there are multiple commands that need to be run to install certs, we should only prompt the user for elevated permissions once.

Reply:
@oxtoacart - about elevation - it's not used in linux at all. the 'elevate' package doesn't support linux.
so the only place that now runs multiple commands doesn't need anything.

regarding changes to the API - how about we get rid of IsInstalled completely?
On Windows/mac - AddAsTrustedRoot will first check for existence, show the popup if needed.
On Linux - it will just install certificates quietly

This way, the code in mitm.go becomes tiny and all the checking is hidden. As far as I can tell, this is the only place where we use IsInstalled